File: android-starter-pack/MADE/Codelab/codelab-dicoding-mynotes/app/src/main/java/com/edsusantoo/bismillah/mynotes/NoteAdapter.java
package com.edsusantoo.bismillah.mynotes;

import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.v7.widget.CardView;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.ArrayList;

import static com.edsusantoo.bismillah.mynotes.db.DatabaseContract.NoteColumns.CONTENT_URI;

public class NoteAdapter extends RecyclerView.Adapter<NoteAdapter.NoteViewHolder> {

    private ArrayList<Note> listNotes = new ArrayList<>();
    private Activity activity;

    public NoteAdapter(Activity activity) {
        this.activity = activity;
    }

    public ArrayList<Note> getListNotes() {
        return listNotes;
    }

    public void setListNotes(ArrayList<Note> listNotes) {
        this.listNotes.clear();
        this.listNotes.addAll(listNotes);
        notifyDataSetChanged();
    }

    public void addItem(Note note) {
        this.listNotes.add(note);
        notifyItemInserted(listNotes.size() - 1);
    }

    public void updateItem(int position, Note note) {
        this.listNotes.set(position, note);
        notifyItemChanged(position, note);
    }

    public void removeItem(int position) {
        this.listNotes.remove(position);
        notifyItemRemoved(position);
        notifyItemRangeChanged(position, listNotes.size());
    }

    @NonNull
    @Override
    public NoteAdapter.NoteViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int i) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_note, parent, false);
        return new NoteViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull NoteAdapter.NoteViewHolder holder, int position) {
        holder.tvTitle.setText(listNotes.get(position).getTitle());
        holder.tvDate.setText(listNotes.get(position).getDate());
        holder.tvDescription.setText(listNotes.get(position).getDescription());
        holder.cvNote.setOnClickListener(new CustomOnItemClickListener(position, new CustomOnItemClickListener.OnItemClickCallback() {
            @Override
            public void onItemClicked(View view, int position) {
                Intent intent = new Intent(activity, NoteAddUpdateActivity.class);
                // Set intent dengan data uri row note by id
                // content://com.dicoding.picodiploma.mynotesapp/note/id
                Uri uri = Uri.parse(CONTENT_URI + "/" + getListNotes().get(position).getId());
                intent.setData(uri);
                intent.putExtra(NoteAddUpdateActivity.EXTRA_POSITION, position);
                intent.putExtra(NoteAddUpdateActivity.EXTRA_NOTE, listNotes.get(position));
                activity.startActivityForResult(intent, NoteAddUpdateActivity.REQUEST_UPDATE);
            }
        }));
    }

    @Override
    public int getItemCount() {
        return listNotes.size();
    }

    class NoteViewHolder extends RecyclerView.ViewHolder {
        final TextView tvTitle, tvDescription, tvDate;
        final CardView cvNote;

        NoteViewHolder(View itemView) {
            super(itemView);
            tvTitle = itemView.findViewById(R.id.tv_item_title);
            tvDescription = itemView.findViewById(R.id.tv_item_description);
            tvDate = itemView.findViewById(R.id.tv_item_date);
            cvNote = itemView.findViewById(R.id.cv_item_note);
        }
    }
}

File: android-starter-pack/AndroidJetpackPro/Codelab/codelab-dicoding-mynote/app/src/main/java/com/edsusantoo/bismillah/mynote/ui/adapter/NoteAdapter.kt
package com.edsusantoo.bismillah.mynote.ui.adapter

import android.app.Activity
import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.cardview.widget.CardView
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.edsusantoo.bismillah.mynote.R
import com.edsusantoo.bismillah.mynote.database.Note
import com.edsusantoo.bismillah.mynote.helper.NoteDiffCallback
import com.edsusantoo.bismillah.mynote.ui.insert.NoteAddUpdateActivity


class NoteAdapter(private val context: Activity) : RecyclerView.Adapter<NoteAdapter.NoteViewHolder>() {

    private val listNote = ArrayList<Note>()

    fun setListNotes(listNote: List<Note>) {
        val diffCallback = NoteDiffCallback(this.listNote, listNote)
        val diffResult = DiffUtil.calculateDiff(diffCallback)

        this.listNote.clear()
        this.listNote.addAll(listNote)
        diffResult.dispatchUpdatesTo(this)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): NoteViewHolder {
        return NoteViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.item_note, parent, false))
    }

    override fun getItemCount(): Int {
        return listNote.size
    }

    override fun onBindViewHolder(holder: NoteViewHolder, position: Int) {
        holder.bind(listNote[position], context)
    }

    class NoteViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        private val tvTitle: TextView = view.findViewById(R.id.tv_item_title)
        private val tvDescription: TextView = view.findViewById(R.id.tv_item_description)
        private val tvDate: TextView = view.findViewById(R.id.tv_item_date)

        private val cvNote: CardView = view.findViewById(R.id.cv_item_note)

        fun bind(note: Note, context: Activity) {
            tvTitle.text = note.title
            tvDate.text = note.date
            tvDescription.text = note.description

            cvNote.setOnClickListener {
                val intent = Intent(context, NoteAddUpdateActivity::class.java)
                intent.putExtra(NoteAddUpdateActivity.EXTRA_POSITION, adapterPosition)
                intent.putExtra(NoteAddUpdateActivity.EXTRA_NOTE, note)
                context.startActivityForResult(intent, NoteAddUpdateActivity.REQUEST_UPDATE)
            }
        }
    }


}